cpp/src/arrow/io/hdfs_internal.cc

```diff
@@ -153,6 +153,7 @@ Result<std::vector<PlatformFilename>> get_potential_libhdfs_paths() {
   ARROW_ASSIGN_OR_RAISE(auto search_paths, MakeFilenameVector({"", "."}));
 
   // Path from environment variable
+  AppendEnvVarFilename("ARROW_LIBHDFS3_DIR", &search_paths);
```

Comment:
this is a hack-ish solution to support libhdfs3, this would try to add one more call to parse from env variable, needs to check the performance